### PR TITLE
[Transform] Modules for Augmentation

### PR DESCRIPTION
**Pull request by mufeili**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/heterograph.py [hunk 1]
```diff
@@ -2614,6 +2614,76 @@ def is_homogeneous(self):
         """
         return len(self.ntypes) == 1 and len(self.etypes) == 1
 
+    def is_bidirected(self, etype=None, eweight=None):
+        """Return whether the graph is a bidirected graph.
+
```
Comment (jermainewang) [inline python/dgl/heterograph.py:2619]:
```
...

A graph is bidirected if for any edge (u, v) in G with weight w, there exists an edge
(v, u) in G with the same weight.
```

Is this how you implement it? What if the graph has parallel edges?

File: python/dgl/transform/module.py [hunk 2]
```diff
@@ -14,11 +14,21 @@
...
 from .. import convert
 from .. import backend as F
+from .. import function as fn
 from . import functional
 
+try:
+    import torch
+    from torch.distributions import Bernoulli
+except ImportError:
+    pass
```
Comment (jermainewang) [inline python/dgl/transform/module.py:30]: Is there a version requirement?

File: python/dgl/transform/module.py [hunk 3]
```diff
@@ -672,3 +690,538 @@ def __call__(self, g):
...
+
+class PPR(BaseTransform):
+    r"""
+
+    Description
+    -----------
+    Apply personalized PageRank (PPR) to an input graph for diffusion, as introduced in
+    `The pagerank citation ranking: Bringing order to the web
+    <http://ilpubs.stanford.edu:8090/422/>`__. A sparsification will be applied to the
+    weighted adjacency matrix after diffusion.
```
Comment (jermainewang) [inline python/dgl/transform/module.py:777]: > A sparsification will be applied to the weighted adjacency matrix after diffusion.

What do you mean by that?

File: python/dgl/transform/module.py [hunk 4]
```diff
@@ -672,3 +690,538 @@ def __call__(self, g):
...
+
+    This module only works for homogeneous graphs.
+
+    Parameters
+    ----------
+    alpha : float, optional
+        Return probability, which commonly lies in :math:`[0.05, 0.2]`.
```
Comment (jermainewang) [inline python/dgl/transform/module.py:784]: Restart probability?

File: python/dgl/transform/module.py (hunk 2 above)
Comment (mufeili) [reply python/dgl/transform/module.py:30]: The version requirement should be handled by [this](https://github.com/dmlc/dgl/blob/master/python/dgl/backend/pytorch/tensor.py#L18).

File: python/dgl/heterograph.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/heterograph.py:2619]: Yes. I've updated the doc and implementation. It now also handles parallel edges with different edge weights.

File: python/dgl/transform/module.py (hunk 3 above)
Comment (mufeili) [reply python/dgl/transform/module.py:778]: Updated.

File: python/dgl/transform/module.py (hunk 4 above)
Comment (mufeili) [reply python/dgl/transform/module.py:785]: Updated.

File: python/dgl/transform/module.py [hunk 5]
```diff
@@ -672,3 +690,538 @@ def __call__(self, g):
...
+    eweight_name : str, optional
+        :attr:`edata` name to retrieve and store edge weights. If it does
+        not exist in an input graph, this module initializes a weight of 1
+        for all edges. The edge weights should be a tensor of shape :math:`(E)`,
+        where E is the number of edges.
+    eps : float, optional
+        The threshold to preserve edges in sparsification after diffusion.
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/transform/module.py:792]: Updated.

File: python/dgl/transform/module.py [hunk 6]
```diff
@@ -672,3 +690,538 @@ def __call__(self, g):
...
+    avg_degree : int, optional
+        The desired average node degree of the sparsified graph. It will only come into
+        effect only if :attr:`eps` is not specified.
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/transform/module.py:795]: Updated.

File: python/dgl/transform/module.py [hunk 7]
```diff
@@ -672,3 +690,538 @@ def __call__(self, g):
...
+# pylint: disable=C0103
+class HeatKernel(BaseTransform):
+    r"""
+
+    Description
+    -----------
+    Apply heat kernel to an input graph for diffusion, as introduced in
+    `Diffusion kernels on graphs and other discrete structures
+    <https://www.ml.cmu.edu/research/dap-papers/kondor-diffusion-kernels.pdf>`__.
+    A sparsification will be applied to the weighted adjacency matrix after diffusion.
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/transform/module.py:870]: GDC introduces the idea of applying sparsification after diffusion as the weighted adjacency matrix can be very dense. PPR and heat kernel can be formulated as two special cases of GDC. PyG implements general GDC, PPR, and heat kernel as a single module GDC, which makes the code and doc less readable. I decided to implement them as separate modules, hence the common use of sparsification.

That says, sparsification is not common at this point beyond the scope of GDC. If it turns out to be common and used in other works, we can add a module for it later.

File: python/dgl/transform/module.py [hunk 8]
```diff
@@ -672,3 +690,538 @@ def __call__(self, g):
...
+    tensor([[11., 12.],
+            [ 9., 10.],
+            [ 7.,  8.]])
+    """
+    def __call__(self, g):
+        for ntype in g.ntypes:
+            nids = F.astype(g.nodes(ntype), F.int64)
+            perm = F.rand_shuffle(nids)
+            for key, feat in g.nodes[ntype].data.items():
+                g.nodes[ntype].data[key] = feat[perm]
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/transform/module.py:1085]: The same permutation is applied to all features if the nodes have multiple features. So I prefer `NodeShuffle` here, which is also used by PyGCL.

File: python/dgl/transform/module.py [hunk 9]
```diff
@@ -672,3 +690,538 @@ def __call__(self, g):
...
+    -------
+
+    >>> import dgl
+    >>> import torch
+    >>> from dgl import NodeShuffle
+
+    >>> transform = NodeShuffle()
+    >>> g = dgl.graph(([0, 1], [1, 2]))
+    >>> g.ndata['h1'] = torch.tensor([[1., 2.], [3., 4.], [5., 6.]])
+    >>> g.ndata['h2'] = torch.tensor([[7., 8.], [9., 10.], [11., 12.]])
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/transform/module.py:1070]: Thanks. I've fixed it.

File: python/dgl/transform/module.py (hunk 2 above)
Comment (jermainewang) [reply python/dgl/transform/module.py:30]: I see. It seems that some of the modules are PyTorch only. How about import torch in the init function?

File: python/dgl/transform/module.py [hunk 10]
```diff
@@ -672,3 +690,538 @@ def __call__(self, g):
...
+            return self.eps
+
+    def __call__(self, g):
+        # Step1: PPR diffusion
+        # (α - 1) A
+        device = g.device
+        eweight = (self.alpha - 1) * g.edata.get(self.eweight_name, F.ones(
+            (g.num_edges(),), F.float32, device))
+        num_nodes = g.num_nodes()
+        mat = F.zeros((num_nodes, num_nodes), F.float32, device)
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/transform/module.py:840]: I think as long as transform is applied before batching, it is fine.

**Review by jermainewang — APPROVED** [review]

**Comment by mufeili** [comment]
@VoVAllen It seems that PyTorch CI still raises an error about PyTorch version like below:

```
Exception: Detected an old version of PyTorch. Please update torch>=1.8.0 for the best experience.
```

**Comment by VoVAllen** [comment]
@mufeili It's on windows. Please contact @BarclayII to update the pytorch version on windows CI machine

**Comment by jermainewang** [comment]
Does it mean that once this is merged, we need to update the minimal required version of PyTorch?

**Comment by mufeili** [comment]
> Does it mean that once this is merged, we need to update the minimal required version of PyTorch?

Yes, 1.8.0

**Pull request merged**